src/libraries/Common/tests/Tests/System/Net/TestPortPoolTests.cs

```diff
@@ -0,0 +1,281 @@
+                            {
+                                Assert.True(currentPorts.TryRemove(assignment.Port, out _));
+                                assignment.Dispose();
+                            }
+                            livingAssignments.Clear();
+                            returnPortsAfter = rnd.Next(returnPortsAfterTicks);
+                            sw.Start();
+                        }
+                    }, i);
+
+                }
+
+                await Task.WhenAll(workItems);
+
+                return RemoteExecutor.SuccessExitCode;
+            }
+
+            RemoteInvokeOptions options = CreateRemoteOptions($"{FirstUnusedPort} {FirstUnusedPort + portRangeLength}");
+            RemoteExecutor.Invoke(RunTest, options).Dispose();
+        }
+
+        [OuterLoop(OuterLoopReason)]
+        [Fact]
+        public void TestSocketIntegration()
+        {
+            static async Task<int> RunTest()
+            {
+                const int levelOfParallelism = 8;
+                const int requestPerThread = 200;
+
+                Task[] workItems = Enumerable.Repeat(Task.Run(() =>
+                {
+                    for (int i = 0; i < requestPerThread; i++)
+                    {
+                        using Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
+                        using PortLease lease = TestPortPool.RentPortAndBindSocket(socket, IPAddress.Loopback);
+
+                        Assert.True(socket.IsBound);
+                        IPEndPoint ep = (IPEndPoint)socket.LocalEndPoint;
+                        Assert.Equal(lease.Port, ep.Port);
+                    }
+                }), levelOfParallelism).ToArray();
+
+                await Task.WhenAll(workItems);
+                return RemoteExecutor.SuccessExitCode;
+            }
+
+            RemoteExecutor.Invoke(RunTest).Dispose();
+        }
+
+        // This test case is intended to detect a potential OS configuration changes on CI machines
+        // that may prevent TestPortPool to operate correctly. The recommended action is to alter TestPortPool range
+        // in those cases.
+        // Although this test is relatively long running because of the external process execution it triggers,
+        // it's better to keep it in the Inner Loop to detect the potential issues fast.
+        [Fact]
+        public void ConfiguredPortRange_DoesNotOverlapWith_OsDynamicPortRange()
+        {
+            PortRange poolRange = TestPortPool.ConfiguredPortRange;
+            var osRange = PortRange.GetDefaultOsDynamicPortRange();
+            string info = $"TestPortPool port range: {poolRange} | OS Dynamic Port Range: {osRange}";
+            _output.WriteLine(info);
+
+            Assert.False(PortRange.AreOverlappingRanges(poolRange, osRange),
+                $"Overlapping port ranges may prevent correct test execution! {info}" );
+        }
```

Comment:
Instead of introducing complicated logic to predict a port range that is likely to be free on a given execution environment, I have added a test case to detect conflicts early. We should be good enough with this IMO.